### PR TITLE
Stabilize `Utf8Chunks`

### PR DESCRIPTION
**Pull request by dtolnay**
Pending FCP in https://github.com/rust-lang/rust/issues/99543.

This PR includes the proposed modification in https://github.com/rust-lang/libs-team/issues/190 as agreed in https://github.com/rust-lang/rust/issues/99543#issuecomment-2050406568.

### CODE REVIEW

**Comment by dtolnay** [comment]
@bors try

I'd like to try crater to find out whether anything is using bstr's `Utf8Chunk::incomplete` (https://github.com/rust-lang/rust/issues/99543#issuecomment-2056725555). If it is, this PR would break it by making libcore's `utf8_chunks()` take precedence over bstr's ByteSlice extension trait. If there is widespread breakage, we need to discuss adding `incomplete` before stabilizing this feature, and finding out about so many uses would help justify doing so. If there is no breakage, then it's a signal we do not need to add `incomplete` because nobody has real-world uses for it.

**Comment by dtolnay** [comment]
@craterbot check

**Comment by joboet** [comment]
That doesn't look too bad! As far as I can tell, `janetrs` is the only regression, the other ones are spurious or use it as a dependency. I opened https://github.com/GrayJack/janetrs/pull/10 to make it resilient against this.

**Review by joboet — CHANGES_REQUESTED** [review]

File: library/core/src/str/lossy.rs [hunk 1]
```diff
@@ -6,6 +6,44 @@ use crate::iter::FusedIterator;
...
+    ///             write!(repr, "\\x{:02X}", byte).unwrap();
+    ///         }
+    ///     }
+    ///     repr.push('"');
+    ///     repr
+    /// }
+    ///
+    /// fn main() {
+    ///     println!("{}", cstr_literal(b"\xferris the \xf0\x9f\xa6\x80\x07"));
+    /// }
```
[the remark this answers is not in the document]
Comment (dtolnay) [reply library/core/src/str/lossy.rs]: Done. I added an assert comparing against `stringify!(c"\xFErris the 🦀\u{7}")`, which I think will be more understandable than `"c\"\\xFErris the 🦀\\u{7}\""` or `r#"c"\xFErris the 🦀\u{7}""#`, but let me know what you think.

**Review by joboet — APPROVED** [review]
Looks good!
@bors r+

File: library/core/src/str/lossy.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (joboet) [reply library/core/src/str/lossy.rs]: Using `stringify` for this hadn't crossed my mind at all, but I think it's neat!

**Comment by lqd** [comment]
These are all doc regressions, which seems fine for perf triage purposes imo.

**Comment by Mark-Simulacrum** [comment]
I'm not so convinced. This *should* be a much more minor change IMO -- it might be exposing some bad performance in rustdoc (maybe some bad behavior with inherent impls?) and could be worth digging into further. Not marking as triaged yet, will try to spend some time here this week.

**Pull request merged**